control.go

```diff
@@ -5,40 +5,84 @@
-	ControlTypeBeheraPasswordPolicy = "1.3.6.1.4.1.42.2.27.8.5.1"
-	// ControlTypeVChuPasswordMustChange - https://tools.ietf.org/html/draft-vchu-ldap-pwd-policy-00
-	ControlTypeVChuPasswordMustChange = "2.16.840.1.113730.3.4.4"
-	// ControlTypeVChuPasswordWarning - https://tools.ietf.org/html/draft-vchu-ldap-pwd-policy-00
-	ControlTypeVChuPasswordWarning = "2.16.840.1.113730.3.4.5"
-	// ControlTypeManageDsaIT - https://tools.ietf.org/html/rfc3296
-	ControlTypeManageDsaIT = "2.16.840.1.113730.3.4.2"
-)
+// controlTypeMap maps controls to text descriptions
+var controlTypeMap = make(map[string]controlType)
+var controlTypeMapLock = sync.Mutex{}
 
-// ControlTypeMap maps controls to text descriptions
-var ControlTypeMap = map[string]string{
-	ControlTypePaging:               "Paging",
-	ControlTypeBeheraPasswordPolicy: "Password Policy - Behera Draft",
-	ControlTypeManageDsaIT:          "Manage DSA IT",
+type controlType struct {
+	Description string
+	Control     Control
 }
 
-// Control defines an interface controls provide to encode and describe themselves
+// RegisterControl adds a Control to the map of known Controls.
+func RegisterControl(oid, description string, ctrl Control) {
+	if ctrl == nil {
+		return
+	}
+	controlTypeMapLock.Lock()
+	defer controlTypeMapLock.Unlock()
+	controlTypeMap[oid] = controlType{
+		Control:     ctrl,
+		Description: description,
+	}
+}
+
+// GetControl returns an empty new initialized Control for the
+// given ControlType (OID). Returns nil if the Control has not
+// been registered with RegisterControl() before.
+func GetControl(oid string) Control {
+	controlTypeMapLock.Lock()
+	ctrl, ok := controlTypeMap[oid]
+	controlTypeMapLock.Unlock()
+	if !ok {
+		return nil
+	}
+	// Indirect returns the value that v points to. [...] If v is
+	// not a pointer, Indirect returns v.
+	refVal := reflect.ValueOf(ctrl.Control)
+	indVal := reflect.Indirect(refVal)
+	if refVal == indVal {
+		// we got something like "type ControlManageDsaIT bool" as Control.
+		// reflect.New returns pointer type -> use indirect
+		return reflect.Indirect(reflect.New(refVal.Type())).Interface().(Control)
+	}
+	// got pointer to something as Control
+	return reflect.New(indVal.Type()).Interface().(Control)
+}
+
+// ControlDescription returns the description for the Control as it was
+// passed to RegisterControl().
+func ControlDescription(oid string) string {
+	controlTypeMapLock.Lock()
+	defer controlTypeMapLock.Unlock()
+	ctrl, ok := controlTypeMap[oid]
+	if !ok {
+		return ""
+	}
+	return ctrl.Description
+}
+
+// Control defines an interface controls provide to encode,
+// decode and describe themselves
 type Control interface {
 	// GetControlType returns the OID
 	GetControlType() string
 	// Encode returns the ber packet representation
 	Encode() *ber.Packet
 	// String returns a human-readable description
 	String() string
+	// Decode decodes the Control
+	Decode(bool, *ber.Packet) (Control, error)
```

Comment:
this feels like a different interface, maybe something like
```
type ControlDecoder interface {
  DecodeControl(oid string, criticality bool, value *ber.Packet) (Control, error)
}
```

Reply:
Agreed - @vetinari do you want to close this PR and optionally make a new PR with a new and separate interface?